### ISSUE TITLE
adding header to cc_library causes all source files to be rebuilt

### ISSUE THREAD

**Issue by crorvick** [bazelbuild/bazel#8319]
### Description of the problem / feature request:

Adding a header to a `cc_library` causes a full rebuild of that library, even if that header is added to `srcs` and no source files include that header yet.

### Feature requests: what underlying problem are you trying to solve with this feature?

Do not rebuild source files that do not depend on the new header.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

```
$ git clone https://github.com/crorvick/bazel-hello.git
Cloning into 'bazel-hello'...
remote: Enumerating objects: 6, done.
remote: Counting objects: 100% (6/6), done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 38 (delta 0), reused 4 (delta 0), pack-reused 32
Unpacking objects: 100% (38/38), done.
$ cd bazel-hello
$ git co 60dfc8f25eee
HEAD is now at 60dfc8f use library for printing
$ bazel build //...
Starting local Bazel server and connecting to it...
INFO: Analysed 5 targets (22 packages loaded, 471 targets configured).
INFO: Found 5 targets...
INFO: Elapsed time: 2.425s, Critical Path: 0.02s
INFO: 0 processes.
INFO: Build completed successfully, 1 total action
$ git checkout 494f76d571da
HEAD is now at 494f76d add unused private header
$ bazel build -s //...
INFO: Analysed 5 targets (1 packages loaded, 6 targets configured).
INFO: Found 5 targets...
SUBCOMMAND: # //src/util:print [action 'Compiling src/util/print.cc']
(cd /home/crorvick/.cache/bazel/_bazel_crorvick/69d6d6c4fc40b0136e835715b0adc69f/execroot/__main__ && \
  exec env - \
    PATH=/home/crorvick/.local/bin:/home/crorvick/bin:/usr/local/bin:/bin:/usr/bin \
    PWD=/proc/self/cwd \
  /bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer '-std=c++0x' -MD -MF bazel-out/k8-fastbuild/bin/src/util/_objs/print/print.pic.d '-frandom-seed=bazel-out/k8-fastbuild/bin/src/util/_objs/print/print.pic.o' -fPIC -iquote . -iquote bazel-out/k8-fastbuild/genfiles -iquote bazel-out/k8-fastbuild/bin -fno-canonical-system-headers -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c src/util/print.cc -o bazel-out/k8-fastbuild/bin/src/util/_objs/print/print.pic.o)
SUBCOMMAND: # //src/hello:hello [action 'Compiling src/hello/hello.cc']
(cd /home/crorvick/.cache/bazel/_bazel_crorvick/69d6d6c4fc40b0136e835715b0adc69f/execroot/__main__ && \
  exec env - \
    PATH=/home/crorvick/.local/bin:/home/crorvick/bin:/usr/local/bin:/bin:/usr/bin \
    PWD=/proc/self/cwd \
  /bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer '-std=c++0x' -MD -MF bazel-out/k8-fastbuild/bin/src/hello/_objs/hello/hello.pic.d '-frandom-seed=bazel-out/k8-fastbuild/bin/src/hello/_objs/hello/hello.pic.o' -fPIC -iquote . -iquote bazel-out/k8-fastbuild/genfiles -iquote bazel-out/k8-fastbuild/bin -iquote external/bazel_tools -iquote bazel-out/k8-fastbuild/genfiles/external/bazel_tools -iquote bazel-out/k8-fastbuild/bin/external/bazel_tools -fno-canonical-system-headers -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c bazel-out/k8-fastbuild/genfiles/src/hello/hello.cc -o bazel-out/k8-fastbuild/bin/src/hello/_objs/hello/hello.pic.o)
INFO: Elapsed time: 0.358s, Critical Path: 0.16s
INFO: 2 processes: 2 linux-sandbox.
INFO: Build completed successfully, 3 total actions
```

### What operating system are you running Bazel on?

CentOS 7.5.1804

### What's the output of `bazel info release`?

```
$ bazel version
Build label: 0.23.2- (@non-git)
Build target: bazel-out/k8-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Thu May 9 03:57:19 2019 (1557374239)
Build timestamp: 1557374239
Build timestamp as int: 1557374239
```

### If `bazel info release` returns "development version" or "(@non-git)", tell us how you built Bazel.

I ran `compile.sh` after exploding the downloaded zip file.

**Comment by oquenchil** [comment]
This doesn't fit Bazel's model. If you add the header, that will retrigger all actions that have this header as an input. There is no way around that. Even with include scanning (which Bazel doesn't have) this would still retrigger the action.

I can't imagine a PR that achieves this without adding quite a bit of complexity. If you manage to do it more or less cleanly, I can still take a look though but I can't promise it will be merged. If I'm wrong about the complexity, then I agree that the change would be very beneficial.

**Comment by oquenchil** [comment]
The documentation does not say include scanning is enabled in Bazel. I can guarantee that it's not enabled and we don't have plans to enable it. One of the reasons why it is not enabled is because include scanning uses a lot of heuristics that assume a strict header layout in packages, just as it's done internally at Google. When the headers do not comply with the restrictions, there are many hard to track errors. So for now the consensus is that it does not make sense to hook up include scanning in Bazel. This might change in the future.

The reason why you see mentions of include scanning in the codebase is because most of the code is shared with the internal version.

Now with respect to the Bazel model comment. Include scanning won't solve the original issue here. Even in the internal version (where include scanning is enabled), adding a new header to the cc_library will retrigger compilation of all sources. In Bazel we can distinguish two phases (analysis and execution), it is not until after execution that we know whether a header was needed.

Besides, I would imagine that adding new headers to a cc_library happens magnitudes less than actually editing files, so I'm not even sure we should be trying to optimize here.

This is not a bug. I will close the issue now but feel free to continue the discussion.

**Comment by obruns** [comment]
I think this bug *could* be about two things; either
* include scanning or
* depfile parsing

I am not interested in include scanning because I also see the problems and intricacies it has.

You do not mention depfiles, though, and AFAICS almost everything is in place as it should be according to the bullet point that follows immediately after [the recent documentation change](https://github.com/bazelbuild/bazel/commit/1df4c71a9cf52f2dd24c4a10da75bb88631e7fde#diff-bf6400d90d2ee8dba23eefe84a542d13dfadcbb6c972d2d4d35ad59ef7cf9376L1068-L1072):

> An action may realize that some files were not used during its execution. In C++, this is called ".d files": the compiler tells which header files were used after the fact, and in order to avoid the embarrassment of having worse incrementality than Make, Bazel makes use of this fact. This offers a better estimate than the include scanner because it relies on the compiler.

Bazel already stores that information in the action cache as can be seen from the following dump (using the example given by OP):

```
bazel dump --action_cache
...
4, bazel-out/k8-fastbuild/bin/src/util/_objs/print/print.pic.o:
      actionKey = 3ba91468fc56704abbcf32266d7f8a366380fd4e4cf32b10cbc485873c573d8b
      usedClientEnvKey = dfbe34d21e5befc7799dffc18af4490647c5334eb90308e6def458364d3a6a30
      digestKey = 05da8e70f939f52c924cfc8497e1ba8290208c09dede99939d49b5438e144acc
      bazel-out/k8-fastbuild/internal/_middlemen/_S_Ssrc_Sutil_Cprint-BazelCppSemantics_build_arch_k8-fastbuild
      bazel-out/k8-opt-exec-2B5CBBC6/internal/_middlemen/external_Slocal_Uconfig_Ucc_Ccompiler_Udeps
      external/bazel_tools/tools/cpp/grep-includes.sh
      src/util/print.cc
      src/util/print.h
                                 <<<< Would expect src/util/foo.h here if depfile parse result wasn't stored.


      packed_len = 137
...
```

I created a [PR](https://github.com/obruns/bazel/pull/1) against my own repo clone because I consider this an unmergeable proof of concept but I still wanted to give you the opportunity to comment. This shows how CcCompilationContexts can be retrofitted with the information already available in the action cache.

Towards a proper solution:
IIUC, when creating CppCompileActions and CcCompilationContexts in the analysis phase, the action cache should be taken into account. With that, we wouldn't have to tweak `declaredIncludeSrcs` after the fact and could keep all members of CcCompilationContext immutable. Also, we wouldn't break all other languages (which my proof of concept probably does).

Now that include scanning is out of the picture: What are the chances of getting a change that introduces an ActionCache-aware analysis phase merged?

**Comment by obruns** [comment]
Ping @oquenchil , @lberki. Can you please comment on my previous post?

Should you decide that Bazel continues to behave as-is (no feedback from the updated ActionCache into the next analysis phase), I suggest to point this out explicitly, e.g. after

> An action may realize that some files were not used during its execution. In C++, this is called ".d files": the compiler tells which header files were used after the fact, and in order to avoid the embarrassment of having worse incrementality than Make, Bazel makes use of this fact. This offers a better estimate than the include scanner because it relies on the compiler.

add

> Note that Bazel has worse incrementality than Ninja because Bazel invalidates the action key if action inputs changed (both explicit and implicit via `glob()`) regardless of the newly added inputs being used or not.

Also, as someone who has fallen into this pitfall I'd prefer such information to be placed much more prominently than in `CODEBASE.md`, e.g. on a comparison page.

**Comment by lberki** [comment]
I think feedback from the execution phase from the analysis phase won't happen because it's so contrary to how Bazel "thinks". I think this use case should not work, but I'm surprised that it doesn't work. Naively, how I think this should work is as follows:

1. The first instance of the action is executed, parses the .d file, puts the entries from the .d file in the action cache
2. A header is added to the `srcs` of whatever rule creates the action
3. That rule is re-analyzes, the `CppCompileAction` is re-created and as far as Bazel is concerned, it hasn't run input discovery or .d file parsing yet
4. The action executes and the entry from (1) is found in the action cache and it's a cache hit.

Why does (4) not happen? Mind you, I think caching the action in (4) would be incorrect because it's possible that e.g.:

1. An action has `-Ifoo -Ibar` on its command line and `#include "x.h"` in the source code
2. The first version of the action `bar/x.h` (but not `foo/x.h`) on its inputs
3. The action runs, resolves `"x.h"` to `bar/x.h`, that's what gets recorded in the .d file
4. `bar/x.h` is added to the inputs of the action, the action is re-analyzed (but its entry in the action cache remains)
5. The action cache sees that the only input to the action is `bar/x.h`, which is unchanged, and thus it's a cache hit

Here (5) is wrong, because `#include "x.h"` should resolve to `foo/x.h`.

So I think it would be nice to handle this case better, but not at the cost of being incorrect in the above scenario. But why do we not reuse the action cache entry? I thought that it works _only_ based on the command line and the set of inputs, which would result in a cache hit.

**Comment by lberki** [comment]
My guess: it appears that the action key of `CppCompileAction` includes `getDeclaredIncludeSrcs()` (as seen in `CppCompileAction.computeKey()`, which would have this exact effect.

**Comment by obruns** [comment]
Thanks for providing these insights @lberki.

`getDeclaredIncludeSrcs()` is indeed the troublemaker. I have extended `src/test/shell/integration/cpp_test.sh` with a couple more test cases (including one that mimics your `-I` order example (steps 1-5) and did not hit any surprises.

The updated action cache is fine - it contains the used headers as reported by the compiler in the depfiles.

The publicly visible history of `CppCompileAction.java` starts at d08b27fa9701fecfdb69e1b0d1ac2459efc2129b and the declared include srcs have been part of the action key ever since. Can you identify (valid) reasons for making the declared include sources part of the action key?

**Comment by obruns** [comment]
`getDeclaredIncludeSrcs()` must still be used if `!discoversInputs()`, I guess?

**Comment by lberki** [comment]
I'm not sure if this answers your question, but if an action does not discover inputs, there is no need to distinguish between inputs and declared inputs and since the inputs already go into the cache key, there is no need to call `getDeclaredIncludeSrcs()` in `computeKey()`.

**Comment by lberki** [comment]
Argh, I should have been more clear :( Re-reading the above thread, I wasn't.

Let me try again: it's true that there is no need to add `getDeclaredIncludeSrcs()` to the key of the action in *Bazel* because include scanning is hard-disabled there, but there is a good reason to add it in *Blaze* (our internal version): there, we do have include scaning and thus we are vulnerable to the correctness issue described [above](https://github.com/bazelbuild/bazel/issues/8319#issuecomment-861200810).

I'll follow up with next steps on  https://github.com/bazelbuild/bazel/pull/13592 i

**Comment by Yannic** [comment]
I'm aware that I'm probably way late to the party, but I don't think avoiding recompilation is possible at all (that is, at least running include scanning and re-computing the cache key).

At least clang has a builtin preprocessor function for checking whether an include exists ([__has_include](https://clang.llvm.org/docs/LanguageExtensions.html#has-include)), so code may behave differently depending on whether a header exists or not. Arguably, most code does not (and should not!) use that feature, but there are some legitimate reasons to do so.

Consider the following target:

```starlark
# foo/BUILD

cc_library(
    name = "foo",
    srcs = ["foo.cc"],
)
```

```c++
// foo/foo.h

if __has_include("foo/foo.h")
#error thou shall not exist!
#endif  // __has_include("foo.h")
```

Building `//foo` as specified above obviously works, `foo/foo.h` does not exist in the exec root (caveat: with sandboxing / RE).

If, however, we add `hdrs = ["foo.h"]` to the target (or add `foo.h` to `srcs`), the build starts to fail as expected. If we do not re-compile when new (previously unincluded) headers are added, we no longer guarantee that a build is indeed correct w.r.t. incremental builds as the outcome differs depending on whether you got a cache hit for `//foo` or not.

See #13592 for a repro case of above scenario that works today but would break with the PR.

The only way to solve this is with include scanning (and I should probably check whether there is a test-case for above example).

**Comment by obruns** [comment]
Thank you for pointing out this corner case, @Yannic. I totally agree with you that include scanning is required to properly support efficient C++ builds in Bazel. There are at least three things speaking against the depfile-based approach:

* When running remote execution, the first build cannot be efficient because the action cache hasn't been filled yet with the results provided by the compiler via the depfile.
* I don't see how a Starlark-ified rules_cpp could possibly be implemented with depfiles in mind. When the Bazel core is supposed to be language-agnostic, depfiles cannot be considered. And the first item is strong enough to not consider depfiles further.
* The case presented by you.


Interestingly, that case also fails on CMake+Ninja. I've converted your snippet from [#13592 (comment)](https://github.com/bazelbuild/bazel/pull/13592#issuecomment-976694722) to CMake. As you can see from the output, Ninja correctly calls CMake because `unused.h` got added to `b` in `CMakeLists.txt`. However, neither CMake nor Ninja care about reproducible builds the way Bazel does. As a result, the `build.ninja` written by CMake is unchanged and Ninja decides that there is nothing to do.


Despite all that, I am probably going to continue with a depfile-enabled fork. We are migrating from multiple Scons-, Make-, Ant- and Eclipse-based build systems to Bazel. We need to cut some corners to make that happen while still keeping the legacy around. Doing the right thing has to be postponed to when we no longer have to maintain that cruft.
In the meantime I need Bazel's remote caching and Ninja's depfile-parsing to have sufficiently fast incremental builds despite the presence of one massive directory tree of headers. This is what the fork is for.

I tried #13871 on our code base but we have relative filesystem walks which I don't want to touch as long as SCons and Make are around:

```cpp
#include "../../../../look/ma/this/is/so/cool/I/can/walk/out/of/my/directory/../BOOM.hpp"
```

----

Your snipped converted to CMake+Ninja

```sh
function test_ninja_compile_on_header_existence_check() {
  CMAKE=`which cmake`
  local -r pkg=$FUNCNAME
  mkdir -p $pkg
  cat > $pkg/CMakeLists.txt <<'EOF'
cmake_minimum_required(VERSION 3.13)
project(header_existence_check LANGUAGES CXX)

add_executable(a
  a.cc
)

target_link_libraries(a
  PRIVATE  # PRIVATE, PUBLIC, INTERFACE wouldn't matter
    b
)

add_library(b INTERFACE)

target_sources(b INTERFACE
  b.h
)

target_include_directories(b
  INTERFACE
    ${CMAKE_CURRENT_SOURCE_DIR}
)
EOF

  cat > $pkg/a.cc <<EOF
#include "b.h"

int main(void) {
  return 0;
}
EOF

  cat > $pkg/b.h <<EOF
#if __has_include("unused.h")
#error unused.h shall not exist
#endif  // __has_include("unused.h")
EOF

  ${CMAKE} -S $pkg -G Ninja -B build
  ${CMAKE} --build build/ -- -d explain -v || echo "build failed"
  sha512sum build/build.ninja

  cat > $pkg/CMakeLists.txt <<'EOF'
cmake_minimum_required(VERSION 3.13)
project(header_existence_check LANGUAGES CXX)

add_executable(a
  a.cc
)

target_link_libraries(a
  PRIVATE  # PRIVATE, PUBLIC, INTERFACE wouldn't matter
    b
)

add_library(b INTERFACE)

target_sources(b INTERFACE
  b.h
)

target_include_directories(b
  INTERFACE
    ${CMAKE_CURRENT_SOURCE_DIR}
)
EOF

  cat > $pkg/unused.h <<EOF
=== BANANA ===
EOF

  ${CMAKE} --build build/ -- -d explain -v a && echo "unexpected success"
  sha512sum build/build.ninja

  rm --recursive test_ninja_compile_on_header_existence_check
  rm --recursive build
}
```

Output:

```
+ /bin/cmake3 -S test_ninja_compile_on_header_existence_check -G Ninja -B build
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/bazel-testcase/build
+ /bin/cmake3 --build build/ -- -d explain -v
ninja explain: output CMakeFiles/a.dir/a.cc.o older than most recent input /tmp/bazel-testcase/test_ninja_compile_on_header_existence_check/b.h (1637699155221870647 vs 1637699226478577994)
ninja explain: CMakeFiles/a.dir/a.cc.o is dirty
ninja explain: a is dirty
[1/2] /opt/rh/devtoolset-10/root/usr/bin/c++   -I/tmp/bazel-testcase/test_ninja_compile_on_header_existence_check  -MD -MT CMakeFiles/a.dir/a.cc.o -MF CMakeFiles/a.dir/a.cc.o.d -o CMakeFiles/a.dir/a.cc.o -c /tmp/bazel-testcase/test_ninja_compile_on_header_existence_check/a.cc
[2/2] : && /opt/rh/devtoolset-10/root/usr/bin/c++     CMakeFiles/a.dir/a.cc.o  -o a   && :
+ sha512sum build/build.ninja
544ee20a14e76fcd2589e599cc1f577d60edd59e51d5f406ac981b898f778be67f3f11a812a2e1bccd36c16b77162ab30540559865d9084cb0de0ebb2a97782b  build/build.ninja
+ cat
+ cat
+ /bin/cmake3 --build build/ -- -d explain -v a
ninja explain: output build.ninja older than most recent input /tmp/bazel-testcase/test_ninja_compile_on_header_existence_check/CMakeLists.txt (1637699226501577900 vs 1637699226629577375)
[0/1] /usr/bin/cmake3 --regenerate-during-build -S/tmp/bazel-testcase/test_ninja_compile_on_header_existence_check -B/tmp/bazel-testcase/build
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/bazel-testcase/build
ninja: no work to do.
+ echo 'unexpected success'
unexpected success
+ sha512sum build/build.ninja
544ee20a14e76fcd2589e599cc1f577d60edd59e51d5f406ac981b898f778be67f3f11a812a2e1bccd36c16b77162ab30540559865d9084cb0de0ebb2a97782b  build/build.ninja
```
